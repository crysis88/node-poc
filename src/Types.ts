export const TYPES = {
    UserRepository: Symbol('UserRepository'),
    PostRepository: Symbol('PostRepository'),
    CommentRepository: Symbol('CommentRepository'),
    ContextProvider: Symbol('ContextProvider')
}