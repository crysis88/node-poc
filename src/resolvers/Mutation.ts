import { ContextProvider } from "../context";
import { compare } from 'bcryptjs';
import { sign } from 'jsonwebtoken';
import { IUser } from "../models/user.model";
export const Mutation = {
    signUp: (parent: any, args: any, context: ContextProvider) => {
        return context.userRepository.createUser(args.userInput.name, args.userInput.mail, args.userInput.password)
    },
    createPost: async (parent: any, args: any, context: ContextProvider) => {
        const user: IUser = await context.getCurrentUser();
        return context.postRepository.createPost(args.postInput.title, args.postInput.content, user.id, args.postInput.published);
    },
    publishPost: async (parent: any, args: any, context: ContextProvider) => {
        await context.getCurrentUser();
        return context.postRepository.publishPost(args.postId);
    },
    deletePost: async (parent: any, args: any, context: ContextProvider) => {
        const user: IUser = await context.getCurrentUser();
        return context.postRepository.deletePost(args.postId);
    },
    createComment: async (parent: any, args: any, context: ContextProvider) => {
        const user: IUser = await context.getCurrentUser();
        return context.commentRepository.addComment(args.message, args.postId, user.id);
    },
    deleteComment: async (parent: any, args: any, context: ContextProvider) => {
        const user: IUser = await context.getCurrentUser();
        return context.commentRepository.deleteComment(args.commentId);
    },
    login: async (parent: any, args: any, context: ContextProvider) => {
        const user: IUser = await context.userRepository.findUserByEmail(args.mail);
        if (user == null) {
            throw new Error('User not found');
        }
        if (!(await compare(args.password, user.password as string))) {
            throw new Error("Passwords don't match");
        }
        return {
            user,
            token: sign({ userId: user.id }, process.env.TOKEN_SECRET, { expiresIn: "5m" })
        }
    }
}