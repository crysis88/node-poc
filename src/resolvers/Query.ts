import { ContextProvider } from "../context";
import { isValidObjectId } from 'mongoose';

export const Query = {
    postById: async (parent: any, args: any, context: ContextProvider) => {
        if (!isValidObjectId(args.postId)) {
            throw new Error('Post Id is invalid');
        }
        return context.postRepository.findPostById(args.id)
    },

    postByAuthor: async (parent: any, args: any, context: ContextProvider) => {
        const author: String = args.author ? args.author : (await context.getCurrentUser()).id;
        return context.postRepository.findPostByAuthor(author);
    },

    comments: async (parent: any, args: any, context: ContextProvider) => {
        if (!isValidObjectId(args.postId)) {
            throw new Error('Post Id is invalid');
        }
        return context.commentRepository.fetchComments(args.post);
    }
}