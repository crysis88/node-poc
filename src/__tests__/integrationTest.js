import "reflect-metadata";
import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import resolvers from '../resolvers';
import typeDefs from '../type-defs';
import { ApolloServer } from 'apollo-server';
import { Container } from 'inversify';
import { DIModule } from '../../inversify.config';
import { TYPES } from '../Types';

const constructTestServer = (mockAuth) => {
  const container = new Container();
  container.load(DIModule);

  const server = new ApolloServer({
    context: () => {
      const ctx = container.get(TYPES.ContextProvider);
      if (mockAuth) {
        ctx.getCurrentUser = jest.fn(() => { return { _id: 'abc' } });
      }
      return ctx
    },
    typeDefs,
    resolvers,
  });
  return { server, container };
};

// the mocked REST API data


const mockedSignupResponse =
{
  _id: "5e4e61a8dc4d7e6b9026c963",
  name: "sandeep6",
  mail: "s@123456678.com"
}

const mockedCreatePostResponse =
{
  _id: "5e4e8e6b3813b23298b32919",
  title: "Test Post",
  content: "This is atest post",
  published: false
}

const mockedPublishPostResponse =
{
  title: "Test Post",
  published: true
}

const mockedCreateCommentResponse =
{
  _id: "5e4fa29275ce7e528086856a",
  message: "Test Message"
}



const SIGNUP = gql`
  mutation signUp($userInput: NewUserInput!) {
    signUp(userInput: $userInput) {
    _id
      name
      mail
  }
}
`;

const CREATE_POST = gql`
  mutation createPost($postInput: NewPostInput!) {
  createPost(postInput: $postInput) {
    _id
    title
    content
    published
  }
}
`;

const PUBLISH_POST = gql`
  mutation publishPost($postId: String!) {
    publishPost(postId: $postId) {
    title
    published
  }
}
`;

const CREATE_COMMENT = gql`
  mutation createComment($message: String!, $postId: String!) {
    createComment(message: $message, postId: $postId) {
    _id
    message
  }
}
`;

describe('mutations', () => {
  it('User signup', async () => {
    const { server, container } = constructTestServer();
    const userRepo = container.get(TYPES.UserRepository);
    userRepo.createUser = jest.fn(() => mockedSignupResponse);
    const { mutate } = createTestClient(server);
    const res = await mutate({ mutation: SIGNUP, variables: { userInput: { name: "sandeep7", mail: "s@1234566789.com", password: "sandep" } } });
    expect(res.data.signUp._id).toBe("5e4e61a8dc4d7e6b9026c963");
    expect(res).toMatchSnapshot();
  });

  it('create post', async () => {
    const { server, container } = constructTestServer(true);
    const postRepo = container.get(TYPES.PostRepository);
    postRepo.createPost = jest.fn(() => mockedCreatePostResponse);
    const { mutate } = createTestClient(server);
    const res = await mutate({ mutation: CREATE_POST, variables: { postInput: { title: "Test Post", content: "This is atest post", published: false } } });
    expect(res.data.createPost.title).toBe("Test Post");
    expect(res).toMatchSnapshot();
  });

  it('Publish post', async () => {
    const { server, container } = constructTestServer(true);
    const postRepo = container.get(TYPES.PostRepository);
    postRepo.publishPost = jest.fn(() => mockedPublishPostResponse);
    const { mutate } = createTestClient(server);
    const res = await mutate({ mutation: PUBLISH_POST, variables: { postId: "5e4e8e6b3813b23298b32919" } });
    expect(res.data.publishPost.published).toBe(true);
    expect(res).toMatchSnapshot();
  });

  it('create comment', async () => {
    const { server, container } = constructTestServer(true);
    const commentRepo = container.get(TYPES.CommentRepository);
    commentRepo.addComment = jest.fn(() => mockedCreateCommentResponse);
    const { mutate } = createTestClient(server);
    const res = await mutate({ mutation: CREATE_COMMENT, variables: { message: "Test Message",postId: "5e4e8e6b3813b23298b32919"  } });
    expect(res.data.createComment.message).toBe("Test Message");
    expect(res).toMatchSnapshot();
  });
  

});
