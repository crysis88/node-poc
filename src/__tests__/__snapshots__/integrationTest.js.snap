// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations Publish post 1`] = `
Object {
  "data": Object {
    "publishPost": Object {
      "published": true,
      "title": "Test Post",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`mutations User signup 1`] = `
Object {
  "data": Object {
    "signUp": Object {
      "_id": "5e4e61a8dc4d7e6b9026c963",
      "mail": "s@123456678.com",
      "name": "sandeep6",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`mutations create comment 1`] = `
Object {
  "data": Object {
    "createComment": Object {
      "_id": "5e4fa29275ce7e528086856a",
      "message": "Test Message",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`mutations create post 1`] = `
Object {
  "data": Object {
    "createPost": Object {
      "_id": "5e4e8e6b3813b23298b32919",
      "content": "This is atest post",
      "published": false,
      "title": "Test Post",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
